Do not initialize config on spack compiler list

haampie opened the pull request:
`spack compiler list` should show the current config and not write new compilers to config.

haampie committed:
Do not initialize config on spack compiler list

msimberg commented:
Thanks @haampie! I don't know if this already fixes the following, but I noticed that doing `spack spec <spec>` also populates the list of compilers if empty.

alalazo commented:
@msimberg I personally agree with `spack compiler list`, but I think going further than that requires some more discussion. Reasons for that:
1. We have been advertizing this behavior [in the docs](https://spack.readthedocs.io/en/latest/getting_started.html#compiler-configuration) since ever so I think there might be people depending on it (" Spack searches for compilers on your machine automatically the first time it is run")
2. There are other places where we automate configuration setting, if none is present, and I think we should be consistent. I would found it really strange if Spack looks automatically for e.g. core compilers for Lmod configuration but doesn't look for compilers on the platform

msimberg commented:
> @msimberg I personally agree with `spack compiler list`, but I think going further than that requires some more discussion. Reasons for that:
> 
>     1. We have been advertizing this behavior [in the docs](https://spack.readthedocs.io/en/latest/getting_started.html#compiler-configuration) since ever so I think there might be people depending on it (" Spack searches for compilers on your machine automatically the first time it is run")
> 
>     2. There are other places where we automate configuration setting, if none is present, and I think we should be consistent. I would found it really strange if Spack looks automatically for e.g. core compilers for Lmod configuration but doesn't look for compilers on the platform

Thanks @alalazo for giving a bit of context. I wanted to open https://github.com/spack/spack/issues/28038 since I really wasn't sure if it was intended behaviour or not. It was surprising behaviour to me, since it definitely was not "the first time" I ran spack. The context was that I was planning to change to a new system-provided config file, and first wanted to check that I didn't have any leftover compilers defined elsewhere before I changed to the new config.

Arguably "the first time it is run" can have quite a few different definitions and does not need to be equivalent to "running `spack compiler list` if there are no configured compilers".

I'm obviously personally happy with this PR, but I don't have any demands that this is the right fix and lack the bigger picture.

tgamblin reviewed (changes requested):
I think this is a good idea -- for 0.18.0 (though we'll hopefully be adding compilers as dependencies for that so this might change quite a bit before the release).

The docs don't actually seem to have anything that says `spack compiler list` should find compilers -- I grepped around and couldn't find it.  They do have some sort of ambiguous mentions of `spack compilers` in `configuration.rst` and `getting_started.rst`.

Two requests:
1. Update `configuration.rst` and `getting_started.rst`.
2. If `spack compiler list`/`spack compilers` finds no compilers, it should prompt the user to run `spack compiler find` (if it's run from a `tty`).  Can you add that?

haampie committed:
Prompt to search for compilers

When `spack compiler list` is run without being restricted to a
particular scope, and no compilers are found, prompt to search for
compilers.

haampie commented on lib/spack/spack/cmd/compiler.py:
```diff
@@ -146,9 +146,21 @@ def compiler_info(args):
+
+    # If no compilers are found, prompt to search for compilers.
+    if args.scope is None and len(compilers) == 0 and sys.stdout.isatty():
+        if not tty.get_yes_or_no("No compilers are configured, search for compilers?"):
```
```suggestion
        if not tty.get_yes_or_no("No compilers available. Search for compilers?"):
```

haampie committed:
Update lib/spack/spack/cmd/compiler.py

haampie commented:
With the last commits:

```
$ SPACK_USER_CONFIG_PATH=$(mktemp -d) spack compiler list
==> No compilers available. Search for compilers? [y/n] n
```

```
$ SPACK_USER_CONFIG_PATH=$(mktemp -d) spack compiler list
==> No compilers available. Search for compilers? [y/n] y
==> Available compilers
-- clang ubuntu20.04-x86_64 -------------------------------------
clang@11.0.0  clang@10.0.0  clang@9.0.1  clang@8.0.1  clang@7.0.1

-- gcc ubuntu20.04-x86_64 ---------------------------------------
gcc@10.3.0  gcc@9.3.0  gcc@8.4.0  gcc@7.5.0
```

```
$ SPACK_USER_CONFIG_PATH=$(mktemp -d) ~/spack/bin/spack compiler list | tee x
==> No compilers available. Search for compilers? [y/n] y
==> Available compilers
-- clang ubuntu20.04-x86_64 -------------------------------------
clang@11.0.0
clang@10.0.0
clang@9.0.1
clang@8.0.1
clang@7.0.1

-- gcc ubuntu20.04-x86_64 ---------------------------------------
gcc@10.3.0
gcc@9.3.0
gcc@8.4.0
gcc@7.5.0
```

```
$ echo "hello world" | env SPACK_USER_CONFIG_PATH=$(mktemp -d) spack compiler list
==> No compilers available
```

```
$ SPACK_USER_CONFIG_PATH=$(mktemp -d) spack compiler list --scope=system
==> No compilers available
```

haampie commented:
@tgamblin I'm not sure what docs sections you were referring to in configuration.rst

haampie committed:
Improve docs

haampie committed:
Check if stdin is a tty

haampie commented:
bump @tgamblin

haampie commented:
bump @tgamblin

haampie committed:
Remove the interactive prompt, leave it at a comment when stdout is a tty

haampie commented:
Okay, the interactive prompt is gone again, it's just showing nothing now when stdout is not a tty:

```console
$ SPACK_USER_CONFIG_PATH=$PWD spack compiler list 
==> No compilers available. Run `spack compiler find` to autodetect compilers
$ SPACK_USER_CONFIG_PATH=$PWD spack compiler list --scope=system
==> No compilers available
$ SPACK_USER_CONFIG_PATH=$PWD spack compiler list | tee
$ echo $?
0
```

and there's a test that otherwise fails on develop.
